### PR TITLE
fix(executor, frontend): `DynamicFilter` type casting of predicate input, NULL RHS means flush delete 

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
`DynamicFilter` type casting of predicate input, NULL RHS means flush delete 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- Type cast fixes: https://github.com/risingwavelabs/risingwave/issues/5771
- No error on RHS not insert, flush deletes instead: https://github.com/risingwavelabs/risingwave/issues/5773

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/tpch.yaml [hunk 1]
```diff
@@ -3409,10 +3409,11 @@
...
     | └─LogicalScan { table: nation, output_columns: [nation.n_nationkey], required_columns: [n_nationkey, n_name], predicate: (nation.n_name = 'KENYA':Varchar) }
-    └─LogicalJoin { type: Inner, on: IsNotDistinctFrom(partsupp.ps_partkey, partsupp.ps_partkey) AND IsNotDistinctFrom(partsupp.ps_suppkey, partsupp.ps_suppkey) AND (partsupp.ps_availqty > (0.5:Decimal * sum(lineitem.l_quantity))), output: [partsupp.ps_suppkey] }
-      ├─LogicalJoin { type: LeftSemi, on: (partsupp.ps_partkey = part.p_partkey), output: all }
-      | ├─LogicalScan { table: partsupp, columns: [partsupp.ps_partkey, partsupp.ps_suppkey, partsupp.ps_availqty] }
-      | └─LogicalScan { table: part, output_columns: [part.p_partkey], required_columns: [p_partkey, p_name], predicate: Like(part.p_name, 'forest%':Varchar) }
+    └─LogicalJoin { type: Inner, on: IsNotDistinctFrom(partsupp.ps_partkey, partsupp.ps_partkey) AND IsNotDistinctFrom(partsupp.ps_suppkey, partsupp.ps_suppkey) AND (partsupp.ps_availqty::Decimal > (0.5:Decimal * sum(lineitem.l_quantity))), output: [partsupp.ps_suppkey] }
+      ├─LogicalProject { exprs: [partsupp.ps_partkey, partsupp.ps_suppkey, partsupp.ps_availqty::Decimal] }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/tpch.yaml:3413]: Not sure why the type cast is done twice. Shouldn't it be optimized away? Anw, I think there was an old comment about this ><

**Comment by chenzl25** [comment]
#5882. I think a better way to deal with the right input of dynamic filter is always requiring exact one row. As we know `SimpleAgg` could  and should always return exact one row.

**Comment by jon-chuang** [comment]
> always requiring exact one row

This doesn’t exactly solve the problem. The row in question could still be NULL.

**Comment by jon-chuang** [comment]
Ok, should have addressed review comments.

File: src/stream/src/executor/dynamic_filter.rs [hunk 2]
```diff
@@ -303,25 +302,31 @@ impl<S: StateStore> DynamicFilterExecutor<S> {
...
                             }
                             _ => {
-                                last_is_insert = false;
+                                // To be consistent, there must be an existing `current_epoch_value`
+                                // equivalent to row indicated for
+                                // deletion.
+                                if Some(row.value_at(0))
+                                    != current_epoch_value.as_ref().map(to_datum_ref)
+                                {
+                                    return Err(anyhow::anyhow!(
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply src/stream/src/executor/dynamic_filter.rs:318]: We need the into().

[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/stream/src/executor/dynamic_filter.rs:318]: `risingwave_common::error::bail`

**Pull request merged**